### PR TITLE
Fix lanzou file list and check available domain

### PR DESCRIPTION
**Pull request by jianyun8023**

### CODE REVIEW

File: internal/client/client.go [hunk 1]
```diff
@@ -90,6 +90,12 @@ func (c *Config) baseURL() string {
 	return "http://" + c.Host
 }
 
+func (c *Client) SetHost(host string) *Client {
```
Comment (syhily) [inline internal/client/client.go]: How about `SetDefaultHost`？

File: internal/driver/lanzou.go [hunk 2]
```diff
@@ -31,16 +30,14 @@ func (l *lanzouDriver) Resolve(shareLink string, passcode string) ([]Share, erro
 	if err != nil {
 		return nil, err
 	}
-	if resp.Code != 200 {
-		return nil, fmt.Errorf("parsed faild: %v", resp.Msg)
+	shareList := make([]Share, len(*resp))
+	for i, item := range *resp {
+		shareList[i] = Share{
+			FileName: item.Name,
```
Comment (syhily) [inline internal/driver/lanzou.go:36]: `Size` is required in `Share`.

File: internal/driver/lanzou/common.go [hunk 3]
```diff
@@ -11,10 +12,33 @@ type Drive struct {
...
 func hostname() string {
-	// We may add a roundrobin list in the future.
-	return "lanzoux.com"
+	return availableDomains[0]
+}
+
+func checkOrSwitchDomain(c *client.Client) (err error) {
+	head, err := c.R().Head("/")
+	if head.IsError() || err != nil {
+		if len(availableDomains) == 1 {
```
Comment (syhily) [inline internal/driver/lanzou/common.go]:
```suggestion
		if len(availableDomains) == 0 {
```

File: internal/driver/lanzou/common.go [hunk 4]
```diff
@@ -11,10 +12,33 @@ type Drive struct {
...
+			return fmt.Errorf("no lanzou domains available")
+		}
+		availableDomains = availableDomains[1:]
+		c.SetHost(hostname())
+		err = checkOrSwitchDomain(c)
```
Comment (syhily) [inline internal/driver/lanzou/common.go]:
```suggestion
		return checkOrSwitchDomain(c)
```

File: internal/driver/lanzou/common.go [hunk 5]
```diff
@@ -11,10 +12,33 @@ type Drive struct {
...
+)
+
+// hostname is used to return a lanzou host.
 func hostname() string {
-	// We may add a roundrobin list in the future.
-	return "lanzoux.com"
+	return availableDomains[0]
+}
+
+func checkOrSwitchDomain(c *client.Client) (err error) {
```
Comment (syhily) [inline internal/driver/lanzou/common.go]:
```suggestion
func checkOrSwitchDomain(c *client.Client) error {
```

File: internal/driver/lanzou/common.go [hunk 6]
```diff
@@ -11,10 +12,33 @@ type Drive struct {
...
+	head, err := c.R().Head("/")
+	if head.IsError() || err != nil {
+		if len(availableDomains) == 1 {
+			return fmt.Errorf("no lanzou domains available")
+		}
+		availableDomains = availableDomains[1:]
+		c.SetHost(hostname())
+		err = checkOrSwitchDomain(c)
+	}
+	return err
```
Comment (syhily) [inline internal/driver/lanzou/common.go]:
```suggestion
	return nil
```

File: internal/driver/lanzou/common.go [hunk 7]
```diff
@@ -34,6 +58,10 @@ func NewDrive(config *client.Config) (*Drive, error) {
 		SetHeader("Accept-Language", "zh-CN,zh;q=0.9").
 		SetHeader("Referer", cl.BaseURL)
 
+	err = checkOrSwitchDomain(cl)
+	if err != nil {
```
Comment (syhily) [inline internal/driver/lanzou/common.go]:
```suggestion
	if err := checkOrSwitchDomain(cl); err != nil {
```

File: internal/driver/lanzou/share.go [hunk 8]
```diff
@@ -13,13 +13,16 @@ import (
 	"github.com/bookstairs/bookhunter/internal/log"
 )
 
-func (l *Drive) ResolveShareURL(shareURL string, pwd string) (*Response, error) {
+func (l *Drive) ResolveShareURL(shareURL string, pwd string) (*[]ResponseData, error) {
```
Comment (syhily) [inline internal/driver/lanzou/share.go]: Use slice is ok, no need to add pointer.

File: internal/driver/lanzou/share.go [hunk 9]
```diff
@@ -13,13 +13,16 @@ import (
...
 	// 移除url前部的主机
 	rawURL, _ := url.Parse(shareURL)
 	parsedURI := rawURL.RequestURI()
 
 	if l.IsFileURL(shareURL) {
-		return l.resolveFileShareURL(parsedURI, pwd)
+		fileShareURL, err := l.resolveFileShareURL(parsedURI, pwd)
+		return &[]ResponseData{
+			*fileShareURL,
+		}, err
```
Comment (syhily) [inline internal/driver/lanzou/share.go]: `ditto`

File: internal/driver/lanzou/share.go [hunk 10]
```diff
@@ -41,9 +44,11 @@ func (l *Drive) removeNotes(html string) string {
 	return html
 }
 
-func (l *Drive) resolveFileShareURL(parsedURI string, pwd string) (*Response, error) {
-	get, _ := l.client.R().Get(parsedURI)
-
+func (l *Drive) resolveFileShareURL(parsedURI string, pwd string) (*ResponseData, error) {
+	get, err := l.client.R().Get(parsedURI)
```
Comment (syhily) [inline internal/driver/lanzou/share.go]:
```suggestion
	resp, err := l.client.R().Get(parsedURI)
```

Use `resp` for same naming.

File: internal/driver/lanzou/share.go [hunk 11]
```diff
@@ -41,9 +44,11 @@ func (l *Drive) removeNotes(html string) string {
 	return html
 }
 
-func (l *Drive) resolveFileShareURL(parsedURI string, pwd string) (*Response, error) {
-	get, _ := l.client.R().Get(parsedURI)
-
+func (l *Drive) resolveFileShareURL(parsedURI string, pwd string) (*ResponseData, error) {
```
Comment (syhily) [inline internal/driver/lanzou/share.go:48]: `resolveFileShareURL` is too long. How about split it into small `func`?

File: internal/driver/lanzou/share.go [hunk 12]
```diff
@@ -207,7 +215,7 @@ func (l *Drive) extractRegex(reg *regexp.Regexp, str string) string {
 	return ""
 }
 
-func (l *Drive) resolveFileItemShareURL(parsedURI string, pwd string) (*Response, error) {
+func (l *Drive) resolveFileItemShareURL(parsedURI string, pwd string) (*[]ResponseData, error) {
```
Comment (syhily) [inline internal/driver/lanzou/share.go]: `ditto`, use pure slice.

**Review by syhily — CHANGES_REQUESTED** [review]
Thanks for your contribution. The code quality is good. I left some comments and hope you can check them.

File: internal/driver/lanzou/common.go (hunk 3 above)
Comment (jianyun8023) [reply internal/driver/lanzou/common.go]: The next step is to intercept the array. A value of 1 is correct.

**Pull request merged**